src/AppShed/Extensions/StorageBundle/Controller/ApiController.php

```diff
@@ -0,0 +1,444 @@
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Method;
+use Sensio\Bundle\FrameworkExtraBundle\Configuration\Template;
+use Symfony\Component\HttpFoundation\JsonResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+class ApiController extends StorageController
+{
+    /**
+     * @Route("/api/list", name="api_list")
+     * @Method({"GET"})
+     * @Template()
+     */
+    public function listAction(Request $request)
+    {
+        $app = $this->getApp($request);
+        $data['apis'] = $this->getDoctrine()->getRepository('AppShedExtensionsStorageBundle:Api')->findBy(['app' => $app]);
+        $data['appParams'] = $this->getExtensionAppParameters($request);
+        return $data;
+    }
+
+    /**
+     * @Route("/api/new", name="api_new")
+     * @Method({"GET", "POST"})
+     * @Template()
+     */
+    public function newAction(Request $request)
+    {
+        $app = $this->getApp($request);
+        $appParams = $this->getExtensionAppParameters($request);
+
+        $api = new Api();
+        $api->setApp($app);
+        $form = $this->createForm(new ApiType($app), $api, [
+            'action' => $this->generateUrl('api_new', $appParams),
+            'method' => 'POST'
+        ]);
+
+        $form->handleRequest($request);
+
+        if ($form->isValid()) {
+            $em = $this->getDoctrine()->getManager();
+            $em->persist($api);
+            $em->flush();
+            $this->get('session')->getFlashBag()->add(
+                'success',
+                'New API created successfully'
+            );
+            if (in_array($api->getAction(), [Api::ACTION_INSERT])) {
+                return $this->redirect($this->generateUrl('api_list', $appParams));
+            }
+            return $this->redirect($this->generateUrl('api_edit', array_merge($appParams, ['uuid' => $api->getUuid()])));
+        }
+
+        $data = [
+            'api' => $api,
+            'form'   => $form->createView(),
+            'appParams' => $appParams
+        ];
+        return $data;
+    }
+
+    /**
+     * @Route("/api/{uuid}/edit", name="api_edit")
+     * @param Api $api
+     * @param $uuid
+     * @ParamConverter("api", class="AppShed\Extensions\StorageBundle\Entity\Api", options={"uuid"="uuid"})
+     * @Method({"GET", "POST"})
+     * @return array
+     * @Template()
+     */
+    public function editAction(Request $request, Api $api, $uuid)
+    {
+        $app = $this->getApp($request);
+        $appParams = $this->getExtensionAppParameters($request);
+        if ($api->getApp() != $app) {
+            throw new NotFoundHttpException("Api with uuid $uuid not found");
+        }
+
+        $form = $this->createForm(new ApiEditType($api), $api, [
+            'action' => $this->generateUrl('api_edit', array_merge($appParams, ['uuid' => $api->getUuid()])),
+            'method' => 'POST'
+        ]);
+
+        switch ($api->getAction()) {
```

Comment:
You should be doing this inside the `Type` - you pass the `$api` to the constructor, so you can mark the type as a object variable and generate the form differently.